User Notifications Page Filter

Ahmedgagan opened the pull request:
- Adds a select kit to filter notifications on user notifications page as discussed

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/discourse/discourse?pullRequest=9535) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/discourse/discourse?pullRequest=9535) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/discourse/discourse?pullRequest=9535) it.</sub>

eviltrout commented:
Looks like there are still linting errors. Could you resolve those please before having us review?

Ahmedgagan commented:
> Looks like there are still linting errors. Could you resolve those please before having us review?

I can see a few files listed by prettier but not the actual issues. How do I find them?

eviltrout commented:
Hi @Ahmedgagan - you should be able to run prettier locally to fix the files.

Ahmedgagan commented:
> Hi @Ahmedgagan - you should be able to run prettier locally to fix the files.

I've fixed all the linting issues via prettier. I can see a few Qunit tests failing but those are probably unrelated?

eviltrout commented:
@jjaffeux can you review this one? The test failure does not happen locally for me. But I think your knowledge of dropdowns would be useful here.

jjaffeux commented:
Yes will review tomorrow.

jjaffeux commented:
Im sorry this was too hard to review line by line as I disagree with the general idea. I don't think we need all this complexity.

I made you a diff of what I would prefer:
[user-notifications-filter.patch.zip](https://github.com/discourse/discourse/files/4551371/user-notifications-filter.patch.zip)

It would now show like this:
<img width="566" alt="Screenshot 2020-04-29 at 13 29 27" src="https://user-images.githubusercontent.com/339945/80591142-75026f80-8a1d-11ea-936f-20eaf5a1ae9b.png">

It doesn't rely on a custom select-kit component and prevents multiple issues, I let you test it apply it if you don't see any problem with this approach, then I can do another review.

Ahmedgagan commented:
Thanks for your review and code @jjaffeux, I understood the changes specified by you, should I use the above mockup or the one discussed <a href="https://meta.discourse.org/t/notifications-unread-only-filter/37621/25">here </a>

jjaffeux commented:
Didnt know there was a mockup for this :(

Should have been linked in initial description of the PR.

So ok You can try to reuse what I did but overwrite only the header please.

Ahmedgagan commented:
@jjaffeux done as specified, and all issues are solved 🙂, please do have a look.

jjaffeux commented:
Cool, thanks. Will do first thing tomorrow morning.

jjaffeux commented:
Please look at my git diff, I showed you how to make it. The `/select-kit/components/notifications-filter.js.es6` file is still totally incorrect, look at how I do computed, and how I use actions to update value, and not mutate it in place.

Ahmedgagan commented:
I have replaced mutation in place with actions as you specified

jjaffeux commented on app/assets/javascripts/select-kit/components/notifications-filter.js.es6:
```diff
@@ -0,0 +1,25 @@
+
+export default DropdownSelectBoxComponent.extend({
+  classNames: ["notifications-filter"],
+  content: [
```
This is incorrect, I showed the correct way in the diff

jjaffeux commented on app/assets/javascripts/select-kit/components/notifications-filter.js.es6:
```diff
@@ -0,0 +1,25 @@
+  ],
+  isVisible: true,
+  valueProperty: null,
+  nameProperty: "value",
```
You should just use the defaults name and id as shown in my diff

jjaffeux commented on app/assets/javascripts/select-kit/templates/components/notifications-filter/notifications-filter-header.hbs:
```diff
@@ -0,0 +1,7 @@
+<span class="filter-text">
+  Filter By
```
This need to be translated

jjaffeux commented on app/assets/javascripts/select-kit/components/notifications-filter.js.es6:
```diff
@@ -0,0 +1,25 @@
+      value: I18n.t("user.user_notifications.filters.unread")
+    }
+  ],
+  isVisible: true,
```
We don't need this line as show in my diff

jjaffeux commented on app/assets/javascripts/discourse/app/templates/user/notifications-index.hbs:
```diff
@@ -8,8 +8,11 @@
   </div>
 {{/if}}
 
+{{notifications-filter value=filter onChange=(action "filterNotifications")}}
```
only the value should be passed as value not the while filter object

jjaffeux commented on app/assets/javascripts/select-kit/templates/components/notifications-filter/notifications-filter-header.hbs:
```diff
@@ -0,0 +1,7 @@
+  Filter By
+</span>
+<span class="header-text">
+  {{value.value}}
```
this should probably just be label

jjaffeux commented on app/assets/javascripts/discourse/app/controllers/user-notifications.js:
```diff
@@ -8,14 +8,23 @@ export default Controller.extend({
+  @discourseComputed("model.content.length", "filter")
+  hasNotifications(length, filter) {
+    if (filter.type === "read") {
+      return this.model.filterBy("read", true).length > 0;
```
You can probably ditch the `> 0` in both cases

```
if (0) {
  console.log("foo"); // won't be called
}
```

jjaffeux commented on app/assets/javascripts/discourse/app/controllers/user-notifications.js:
```diff
@@ -8,14 +8,23 @@ export default Controller.extend({
   currentPath: readOnly("router._router.currentPath"),
+  filter: {
+    type: "all",
+    value: I18n.t("user.user_notifications.filters.all")
```
value should be a localisable string, it should be kinda unique

Ahmedgagan commented:
@jjaffeux 
You've referred to your diff a couple of times. Thanks for going ahead and sharing it. But I think there's a small problem. Here are the contents of your diff.

```
diff --git a/app/assets/javascripts/discourse/app/templates/user/notifications-index.hbs b/app/assets/javascripts/discourse/app/templates/user/notifications-index.hbs
index 0862699e43..0483811f28 100644
--- a/app/assets/javascripts/discourse/app/templates/user/notifications-index.hbs
+++ b/app/assets/javascripts/discourse/app/templates/user/notifications-index.hbs
@@ -8,7 +8,7 @@
   </div>
 {{/if}}
 
-{{notifications-filter action=(action "filterNotifications")}}
+{{notifications-filter value=filter onChangeFilter=(action "filterNotifications")}}
 <span class="user-notifications-filter-separator"></span>
 
 {{#if hasNotifications}}
diff --git a/app/assets/javascripts/select-kit/components/notifications-filter.js.es6 b/app/assets/javascripts/select-kit/components/notifications-filter.js.es6
deleted file mode 100644
index 65fa59d14d..0000000000
--- a/app/assets/javascripts/select-kit/components/notifications-filter.js.es6
+++ /dev/null
@@ -1,40 +0,0 @@
-import DropdownSelectBoxComponent from "select-kit/components/dropdown-select-box";
-
-export default DropdownSelectBoxComponent.extend({
-  classNames: ["notifications-filter"],
-  content: [
-    {
-      type: "all",
-      value: I18n.t("user.user_notifications.filters.all")
-    },
-    {
-      type: "read",
-      value: I18n.t("user.user_notifications.filters.read")
-    },
-    {
-      type: "unread",
-      value: I18n.t("user.user_notifications.filters.unread")
-    }
-  ],
-  value: {
-    type: "all",
-    value: I18n.t("user.user_notifications.filters.all")
-  },
-  isVisible: true,
-  valueProperty: null,
-
-  modifyComponentForRow() {
-    return "notifications-filter/notifications-filter-row";
-  },
-
-  selectKitOptions: {
-    headerComponent: "notifications-filter/notifications-filter-header"
-  },
-
-  actions: {
-    onChange(filter) {
-      this.set("value", filter);
-      this.attrs.action && this.attrs.action(filter.type);
-    }
-  }
-});
diff --git a/app/assets/javascripts/select-kit/components/notifications-filter/notifications-filter-header.js.es6 b/app/assets/javascripts/select-kit/components/notifications-filter/notifications-filter-header.js.es6
deleted file mode 100644
index 16fdf1a461..0000000000
--- a/app/assets/javascripts/select-kit/components/notifications-filter/notifications-filter-header.js.es6
+++ /dev/null
@@ -1,13 +0,0 @@
-import DropdownSelectBoxHeaderComponent from "select-kit/components/dropdown-select-box/dropdown-select-box-header";
-import discourseComputed from "discourse-common/utils/decorators";
-
-export default DropdownSelectBoxHeaderComponent.extend({
-  layoutName:
-    "select-kit/templates/components/notifications-filter/notifications-filter-header",
-  classNames: ["notifications-filter-header"],
-
-  @discourseComputed("selectKit.isExpanded")
-  caretIcon(isExpanded) {
-    return isExpanded ? "caret-up" : "caret-down";
-  }
-});
diff --git a/app/assets/javascripts/select-kit/templates/components/notifications-filter/notifications-filter-header.hbs b/app/assets/javascripts/select-kit/templates/components/notifications-filter/notifications-filter-header.hbs
deleted file mode 100644
index a7d21ad317..0000000000
--- a/app/assets/javascripts/select-kit/templates/components/notifications-filter/notifications-filter-header.hbs
+++ /dev/null
@@ -1,7 +0,0 @@
-<span class="filter-text">
-  Filter By
-</span>
-<span class="header-text">
-  {{value.value}}
-</span>
-{{d-icon caretIcon class="caret-icon"}}
diff --git a/app/assets/javascripts/select-kit/templates/components/notifications-filter/notifications-filter-row.hbs b/app/assets/javascripts/select-kit/templates/components/notifications-filter/notifications-filter-row.hbs
deleted file mode 100644
index 83583c08c5..0000000000
--- a/app/assets/javascripts/select-kit/templates/components/notifications-filter/notifications-filter-row.hbs
+++ /dev/null
@@ -1,5 +0,0 @@
-<span class="selection-indicator"></span>
-
-<span class="filter-option">
-  {{rowValue.value}}
-</span>
diff --git a/app/assets/stylesheets/common/base/user.scss b/app/assets/stylesheets/common/base/user.scss
index 869635a250..00ae6bda7f 100644
--- a/app/assets/stylesheets/common/base/user.scss
+++ b/app/assets/stylesheets/common/base/user.scss
@@ -70,6 +70,10 @@
     width: 100%;
     border: 0.5px solid $primary-low;
   }
+
+  .user-notifications-filter {
+    margin-bottom: 0.5em;
+  }
 }
 
 .user-main {
diff --git a/app/assets/stylesheets/common/select-kit/notifications-filter.scss b/app/assets/stylesheets/common/select-kit/notifications-filter.scss
deleted file mode 100644
index a2dd03d3ca..0000000000
--- a/app/assets/stylesheets/common/select-kit/notifications-filter.scss
+++ /dev/null
@@ -1,56 +0,0 @@
-.select-kit {
-  &.dropdown-select-box {
-    &.notifications-filter {
-      display: inline-flex;
-      position: relative;
-
-      .select-kit-collection {
-        padding: 5px;
-      }
-
-      .notifications-filter-header {
-        padding: 0 0.5em 0 0.5em;
-        background: none;
-        border: none;
-        justify-content: flex-start;
-        width: auto;
-        height: auto;
-        span.filter-text {
-          margin-right: 15px;
-          color: $primary-medium;
-        }
-
-        span.header-text {
-          color: dark-light-choose($tertiary, $tertiary);
-        }
-
-        .d-icon {
-          color: $primary-medium;
-          opacity: 1;
-          margin: 5px 0 10px 5px;
-          font-size: $font-up-3;
-        }
-
-        &.is-focused,
-        &:hover {
-          background: none;
-          border: none;
-          outline: none;
-        }
-      }
-      .notifications-filter-row {
-        font-weight: bold;
-        padding: 0.5em;
-        font-size: $font-up-1;
-        align-items: center;
-        display: flex;
-        width: auto;
-        height: auto;
-
-        span.filter-option {
-          flex: 1 0 0px;
-        }
-      }
-    }
-  }
-}
```

Are we on the same page on this?

jjaffeux commented:
@Ahmedgagan Oh damn I didn't include new files... Why didn't you say anything first time I told you about this? 😄 Anyways my bad, sorry.

I will make more comments then.

jjaffeux commented:
OK will my last comments, you should be good to go :+1:

jjaffeux commented on app/assets/javascripts/select-kit/components/notifications-filter/notifications-filter-header.js.es6:
```diff
@@ -0,0 +1,13 @@
+import DropdownSelectBoxHeaderComponent from "select-kit/components/dropdown-select-box/dropdown-select-box-header";
```
This file should be .js and not .js.es6

Ahmedgagan commented:
no worries, thanks for helping me out

jjaffeux commented:
Ok :+1: looks better, thanks!

Will try it tomorrow and merge if ok.

Ahmedgagan commented:
thanks.

jjaffeux commented:
Merged thanks!
